driver/src/contracts/batched_auction_data_reader.rs

```diff
@@ -0,0 +1,255 @@
+use super::stablex_auction_element::{StableXAuctionElement, AUCTION_ELEMENT_WIDTH};
+use dfusion_core::models::{AccountState, BatchInformation, Order};
+use std::collections::HashMap;
+use web3::types::{H160, U256};
+
+/// Handles reading of auction data that has been encoded with the smart
+/// contract's `encodeAuctionElement` function.
+pub struct BatchedAuctionDataReader {
+    index: U256,
+    pub account_state: AccountState,
+    /// All orders in the order they were received.
+    pub orders: Vec<Order>,
+    /// The total number of orders per user.
+    pub user_order_counts: HashMap<H160, u16>,
```

Comment:
Maybe it would make sense to only store a user_order Map<H160, Order> on the struct and turn it into a Vec<Order> on demand? Rather than storing two fields? This way we'd have less implicitly semantically coupled fields.

Reply:
Agree that it would be nicer from that perspective but after trying this I think the code is overall better the way it is now because there is no extra work to produce the final order vector and parsing the orders is a bit simpler.

Reply:
I think if we want to keep it this way, then `orders` and `user_order_counts` can't be public, or else that would allow code outside the module to mutate these fields and invalidate the `BatchedAuctionDataReader`. The issue with that is you are attaching implicit requirements to your data and no way of enforcing it with the compiler.

Reply:
Should be fine with just `user_order_counts` private. Mutating `orders` doesn't change correctness.

Reply:
If I add/delete orders from the vec without updating the user_order_counts that's invalid isn't it? If using a Map<H160, Order> is too ugly maybe we can make the fields private and read only?

Reply:
It doesn't change how parsing new orders works and it doesn't change the pagination logic is what I meant, but it doesn't make sense to change the orders. Really I mostly wanted to avoid writing more getters for it but I'll do that now.

Reply:
The issue is that it doesn’t make sense now and in this context to change the orders vector, it may make sense to someone unfamiliar with the code base some at a different time and place. Having these implicit requirements is an easy way to introduce bugs in the future.